src/block-cache.ts

```diff
@@ -0,0 +1,183 @@
+import {
+  createAsyncMiddleware,
+  JsonRpcMiddleware,
+  JsonRpcRequest,
+} from 'json-rpc-engine';
+import * as cacheUtils from './cache-utils';
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires,@typescript-eslint/no-require-imports
+const BlockTracker = require('eth-block-tracker');
+
+// `<nil>` comes from https://github.com/ethereum/go-ethereum/issues/16925
+const emptyValues: (string|null|undefined)[] = [undefined, null, '\u003cnil\u003e'];
+
+interface BlockCacheMiddlewareOptions{
+  blockTracker?: typeof BlockTracker;
+}
+
+export = createBlockCacheMiddleware;
+
+//
+// Cache Strategies
+//
+
+class BlockCacheStrategy {
+
+  private cache: Record<string, Record<string, unknown>>;
+
+  constructor() {
+    this.cache = {};
+  }
+
+  getBlockCacheForPayload(_payload: JsonRpcRequest<string[]>, blockNumberHex: string): Record<string, unknown> {
+    const blockNumber: number = Number.parseInt(blockNumberHex, 16);
+    let blockCache: Record<string, unknown> = this.cache[blockNumber];
+    // create new cache if necesary
+    if (!blockCache) {
+      const newCache: Record<string, unknown> = {};
+      this.cache[blockNumber] = newCache;
+      blockCache = newCache;
+    }
+    return blockCache;
+  }
+
+  async get(payload: JsonRpcRequest<string[]>, requestedBlockNumber: string): Promise<unknown> {
+    // lookup block cache
+    const blockCache: Record<string, unknown> = this.getBlockCacheForPayload(payload, requestedBlockNumber);
+    if (!blockCache) {
+      return undefined;
+    }
+    // lookup payload in block cache
+    const identifier: string|null = cacheUtils.cacheIdentifierForPayload(payload, true);
+    const cached: unknown = identifier ? blockCache[(identifier as string)] : undefined;
+    // may be undefined
+    return cached;
+  }
+
+  async set(payload: JsonRpcRequest<string[]>, requestedBlockNumber: string, result: Record<string, unknown>): Promise<void> {
+    // check if we can cached this result
+    const canCache: boolean = this.canCacheResult(payload, result);
+    if (!canCache) {
+      return;
+    }
+    // set the value in the cache
+    const blockCache: Record<string, unknown> = this.getBlockCacheForPayload(payload, requestedBlockNumber);
+    const identifier: string|null = cacheUtils.cacheIdentifierForPayload(payload, true);
+    blockCache[(identifier as string)] = result;
```

Comment:
I know that these lines preserve existing functionality, but this looks like a bug to me. If `identifier` is `null`, we should return early, just like we do in the `canCache` check above.